src/app/shared/services/dashboard-products.service.ts

```diff
@@ -0,0 +1,55 @@
+import { Injectable } from '@angular/core';
+import { Product } from '../models/product';
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class DashboardProductsService {
+  constructor() {}
+
+  sumQunatityAvailableByLocation(products:Product[]){
+    const productsCounts = products.map((product) => [...product.counts]);
+    const concatedCounts = [].concat.apply([], productsCounts);
+
+    const reducedByLoction = Array.from(
+      concatedCounts.reduce(
+        (m, { location, quantityAvailable }) =>
+          m.set(location, (m.get(location) || 0) + quantityAvailable),
+        new Map()
+      ),
+      ([location, quantityAvailable]) => ({ location, quantityAvailable })
+    );
+
+    return reducedByLoction;
+  }
+
+  concatCounts(products:Product[]){
+    const productsCounts = products.map((product) => [...product.counts]);
+    const concatedCounts = [].concat.apply([], productsCounts);
+    return concatedCounts;
+  }
+
+  removeDuplicatedProductsByLocationAndPrice(products:Product[]){
+    const productsCounts = products.map((product) => [...product.counts]);
+    const concatedCounts = [].concat.apply([], productsCounts);
```

Comment:
what?? why do you need this apply ?

Reply:
Ok, this is a bit crazy, but it's all because when I have the productCounts it is returned as an array of arrays with objects inside (because I want to have access to all counts from all matching products - originally the charts displayed data from one product, but as I explained before I don't understand an idea of product in this project - ex. we have two products with the same name, picture, description and different counts (first one ex. have location:USA, quantityAvailable:20 and price:800, and the second one - location:USA, quantityAvailable:10 and price:900) but chart shows the data only from the first product , so I changed the charts to display data for both products. That's why at the end I have an array with arrays(counts) inside, and I don't know how many of them is inside so I used that solution to be sure that at the end we will get only one array with all possible objects inside.